infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/external/ExternalServerExposer.java

```diff
@@ -60,18 +66,68 @@ public void expose(
       String serviceName,
       ServicePort servicePort,
       Map<String, ServerConfig> externalServers) {
-    Ingress ingress = generateIngress(machineName, serviceName, servicePort, externalServers);
-    k8sEnv.getIngresses().put(ingress.getMetadata().getName(), ingress);
+    List<Ingress> ingresses =
+        generateIngresses(machineName, serviceName, servicePort, externalServers);
+    for (Ingress ingress : ingresses) {
+      k8sEnv.getIngresses().put(ingress.getMetadata().getName(), ingress);
+    }
   }
 
-  private Ingress generateIngress(
+  private List<Ingress> generateIngresses(
       String machineName,
       String serviceName,
       ServicePort servicePort,
       Map<String, ServerConfig> ingressesServers) {
 
+    // all the unique servers need their separate ingresses. all the other servers can get lumped
+    // under a common ingress corresponding to the server and port.
+    Map<String, ServerConfig> nonUniqueServers = new HashMap<>();
```

Comment:
Do you think it's possible to move this logic one level above in `KubernetesServerExposer`, it would allow avoiding duplicating the same logic in K8s and OS infrastructures. Something like the following https://github.com/metlos/che/compare/bug/15283-multiple-plugin-endpoints-on-the-same-port...sleshchenko:proposal-1?expand=1

Reply:
I've done this in the newest version, only slightly differently. Your suggestion used the secured exposer even for non-secured external servers. I extracted the common logic into basically a free-standing function and reused it in both secured and non-secured exposers below the level of `KubernetesServerExposer` such that we don't assume how servers are exposed on that high level.

Could you please re-review the latest version?